fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java

```diff
@@ -2620,58 +2684,93 @@ private PublishResult finishCheckQuorumReplicas(TransactionState transactionStat
-                // Here we only check number, the replica version will be updated in updateCatalogAfterVisible()
-                for (MaterializedIndex index : allIndices) {
-                    for (Tablet tablet : partition.getIndex(index.getId()).getTablets()) {
-                        List<Replica> tabletSuccReplicas = Lists.newArrayList();
-                        List<Replica> tabletWriteFailedReplicas = Lists.newArrayList();
-                        List<Replica> tabletVersionFailedReplicas = Lists.newArrayList();
-                        // TODO always use the visible version because the replica version is not changed
-                        long newVersion = partition.getVisibleVersion() + 1;
-                        for (Replica replica : tablet.getReplicas()) {
-                            List<PublishVersionTask> publishVersionTasks = publishTasks.get(replica.getBackendId());
-                            PublishVersionTask publishVersionTask = null;
-                            if (publishVersionTasks != null) {
+            }
+            long minSubTxnId = subTxnIds.get(0);
+            long minVersion = transactionState.getTableCommitInfoBySubTxnId(minSubTxnId).getIdToPartitionCommitInfo()
+                    .get(partitionId).getVersion();
+            long maxSubTxnId = subTxnIds.get(subTxnIds.size() - 1);
+            long maxVersion = transactionState.getTableCommitInfoBySubTxnId(maxSubTxnId).getIdToPartitionCommitInfo()
+                    .get(partitionId).getVersion();
+            LOG.debug("txn_id={}, partition={}, min_version={}, max_version={}", transactionState.getTransactionId(),
+                    partitionId, minVersion, maxVersion);
+            boolean alterReplicaLoadedTxn = isAlterReplicaLoadedTxn(minVersion, table);
+            // check success replica number for each tablet.
+            // a success replica means:
+            //  1. Not in errorReplicaIds: succeed in both commit and publish phase
+            //  2. last failed version < 0: is a health replica before
+            //  3. version catch up: not with a stale version
+            // Here we only check number, the replica version will be updated in updateCatalogAfterVisible()
+            for (MaterializedIndex index : allIndices) {
+                for (Tablet tablet : partition.getIndex(index.getId()).getTablets()) {
+                    List<Replica> tabletSuccReplicas = Lists.newArrayList();
+                    List<Replica> tabletWriteFailedReplicas = Lists.newArrayList();
+                    List<Replica> tabletVersionFailedReplicas = Lists.newArrayList();
+                    for (Replica replica : tablet.getReplicas()) {
+                        List<PublishVersionTask> publishVersionTasks = publishTasks.get(replica.getBackendId());
+                        List<PublishVersionTask> replicaTasks = new ArrayList<>();
+                        if (publishVersionTasks != null) {
```

Comment:
when publishVersionTasks == null,  replicaTasks will be empty.   But replicaTasks' size should be the same with subTxnIds.